pkg/controller/projects/variables/controller.go

```diff
@@ -0,0 +1,150 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"context"
+
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	"github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
+)
+
+const (
+	errNotVariable  = "managed resource is not a Gitlab variable custom resource"
+	errGetFailed    = "cannot get Gitlab variable"
+	errCreateFailed = "cannot create Gitlab variable"
+	errUpdateFailed = "cannot update Gitlab variable"
+	errDeleteFailed = "cannot delete Gitlab variable"
+)
+
+// SetupVariable adds a controller that reconciles Variables.
+func SetupVariable(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.VariableKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Variable{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.VariableGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newGitlabClientFn: projects.NewVariableClient}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube              client.Client
+	newGitlabClientFn func(cfg clients.Config) projects.VariableClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Variable)
+	if !ok {
+		return nil, errors.New(errNotVariable)
+	}
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err
+	}
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*cfg)}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client projects.VariableClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Variable)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotVariable)
+	}
+	variable, _, err := e.client.GetVariable(*cr.Spec.ForProvider.ProjectID, cr.Spec.ForProvider.Key)
```

Comment:
Hrm, I guess external-name is kind of unintuitive in this context? I imagine it would technically be the variable key?

Reply:
Yeah this was my thinking. There isn't a single unique value to identify a variable (it is Project ID + Key), so putting the Key into the external-name field is just adding another level potential breakage.